controller/server/update_client.ml

```diff
@@ -0,0 +1,93 @@
+open Lwt
+
+(* unparsed semver version *)
+type version = string
+(* local filesystem path *)
+type bundle_path = string
+
+module type UpdateClientIntf = sig
+    (* download bundle from specified url and save it as `version` *)
+    val download : Uri.t -> version -> bundle_path Lwt.t
+
+    (* URL from which the specified version would be downloaded *)
+    val download_url : version -> Uri.t
+
+    (** Get latest version available *)
+    val get_latest_version : unit -> version Lwt.t
+end
+
+module type UpdateClientConfig = sig
+    (* TODO: convert to Uri.t *)
+    val base_url: string
+    val proxy: Uri.t option
+end
+
+let make_config ?proxy base_url : (module UpdateClientConfig) = (module struct
+    let base_url = base_url
+    let proxy = proxy
+end)
+
+let bundle_name = Config.System.bundle_name
+
+let bundle_file_name version =
+  Format.sprintf "%s-%s.raucb" bundle_name version
+
+(* NOTE/TODO: potential simplification:
+
+    UpdateClient does not care and need to know about the possibility of a
+    proxy, since we are interfacing with curl via bindings that spawn a
+    subprocess it would be sufficient to set the appropriate `http_proxy` env
+    variable system-wide (at run time) and let curl figure out whether it needs
+    to use a proxy or not itself.
```

Comment:
After reconsidering, the env variable approach would be messy, but there's a nicer/simpler alternative: write the `proxy` flag value to `curlrc`, which is read on every `curl` invocation (see `curl` man pages on `--config`).

Above you mentioned that:

> the active proxy might change depending on which network service is the current gateway (which might change without the controller's involvement).

These state updates can be managed by listening to `ServicesChanged` / `PropertyChanged` connman signals, which I see are already exposed in the ocaml bindings. I will look into this, but maybe not in the scope of this PR.

Reply:
I realised that the current code is broken already - if the default proxy changes, `UpdateClientConfig.proxy` will never be updated. I started digging and proxy resolution really is all over the place in the code, e.g. the `kiosk` code is [listening to connman proxy updates](https://github.com/dividat/playos/blob/26dc0346d813f4f38cf847e9d0e5134955f8c990/kiosk/kiosk_browser/proxy.py#L135-L150) separately. Consolidating this could simplify a lot of things.

Reply:
Yes, global proxy resolution on Linux is a higgledy-piggledy affair. There is no one way to configure a system-wide proxy. We treat connman as the source of truth, but have to query it whenever needed.

When you have KDE or Gnome, they can instead [provide global settings which browsers are in the habit of picking up](https://chromium.googlesource.com/chromium/src/+/master/docs/linux/proxy_config.md). But we'd still have to teach the controller how to get the same settings.

Our conclusion has been that there is no clean and simple way. I'm happy to be proven wrong.

Reply:
Switched back to dynamic proxy lookups from connman in  UpdateClient and removed the comment for now. Let's tackle in a separate task, also keeping in mind other proxy config lookups (e.g. kiosk).